test/e2e/catalogutil/catalog.go

```diff
@@ -0,0 +1,1537 @@
+	Dependency 3 Error 0: unknown type: int
+*/
+func (depFile *DependenciesFile) AddDependency(enforceDependencyValidation bool, dependencies ...interface{}) error {
+	errorsMap := map[int][]error{}
+
+	addErr := func(argumentIndex int, errs ...error) {
+		if errorList, ok := errorsMap[argumentIndex]; ok {
+			errorList = append(errorList, errs...)
+		} else {
+			errorList = append(errorList, errs...)
+			errorsMap[argumentIndex] = errorList
+		}
+	}
+
+	for depIndex, dependency := range dependencies {
+		discoveredType := ""
+
+		switch dependency.(type) {
+		case registry.GVKDependency:
+			discoveredType = registry.GVKType
+			gvk := dependency.(registry.GVKDependency)
+			errs := gvk.Validate()
+			if enforceDependencyValidation && len(errs) > 0 {
+				addErr(depIndex, errs...)
+				continue
+			}
+		case registry.PackageDependency:
+			discoveredType = registry.PackageType
+			gvk := dependency.(registry.PackageDependency)
+			errs := gvk.Validate()
+			if enforceDependencyValidation && len(errs) > 0 {
+				addErr(depIndex, errs...)
+				continue
+			}
+		default:
+			addErr(depIndex, fmt.Errorf("unknown type: %T", dependency))
+			continue
+		}
+		b, err := json.Marshal(dependency)
+		if err != nil {
+			addErr(depIndex, fmt.Errorf("Unable to marshal %s: %v", discoveredType, err))
+			continue
+		}
+		depFile.Dependencies = append(depFile.Dependencies, registry.Dependency{Type: discoveredType, Value: b})
+	}
+	if len(errorsMap) > 0 {
+		collectedErrorMsg := "Collected errors:\n"
+		for argumentIndex, errorList := range errorsMap {
+			for errorIndex, e := range errorList {
+				collectedErrorMsg += fmt.Sprintf("\tDependency %d Error %d: %s\n", argumentIndex, errorIndex, e.Error())
+			}
+		}
+		return errors.Errorf("%s", collectedErrorMsg)
+	}
+	return nil
+}
+
+// WriteFile writes contents to <destinationDir>/dependencies.yaml if dependencies are present
+func (depFile *DependenciesFile) WriteFile(destinationDir string) error {
+	// if there are no dependencies defined, there's no point in writing out the file
+	if len(depFile.Dependencies) == 0 {
+		return nil
+	}
+
+	yamlBytes, err := yaml.Marshal(depFile)
+	if err != nil {
+		return err
+	}
+	err = ioutil.WriteFile(filepath.Join(destinationDir, "dependencies.yaml"), yamlBytes, os.FileMode(0644))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+/*
+CreateTemporaryCatalog is the main entry point for generating a temporary catalog for use with e2e tests.
+
+Arguments:
+
+• toolsBin: The absolute path to the directory that contains the tools needed to generate a catalog
+
+• catalogEntries: a slice of catalog entries that need to be added into a catalog
+
+• stack: contains arguments needed to generate a catalog for a specific environment
+
+Returns:
+
+error is non nil if anything prevented the catalog from being generated and pushed to the target registry
+*/
+func CreateTemporaryCatalog(toolsBin string, catalogEntries []CatalogEntry, stack Stack) error {
```

Comment:
This is great, and it's frankly impressive that you got this to work. With the move to declarative config based indexes though, I wonder if there is a way to greatly simplify this by simply using config test fixtures.

Reply:
We started this test suite MANY months ago to explicitly test out the `opm` imperative design.  I think we should finish this quickly and then move on to another declarative approach.